src/main/java/jenkins/branch/MultiBranchProject.java

```diff
@@ -201,6 +203,48 @@ public void onLoad(ItemGroup<? extends Item> parent, String name) throws IOExcep
             LOGGER.log(Level.WARNING, "Could not read persisted state, will be recovered on next index.", e);
             state.reset();
         }
+        BranchProjectFactory<P, R> factory = getProjectFactory();
+        // optimize lookup of sources by building a temporary map that is equivalent to getSCMSource(id) in results
+        Map<String,SCMSource> sourceMap = new HashMap<>();
+        for (BranchSource source : sources) {
+            SCMSource s = source.getSource();
+            String id = s.getId();
+            if (!sourceMap.containsKey(id)) { // only the first match should win
+                sourceMap.put(id, s);
+            }
+        }
+        for (P item : getItems()) {
+            if (factory.isProject(item)) {
+                Branch oldBranch = factory.getBranch(item);
+                SCMSource source = sourceMap.get(oldBranch.getSourceId());
+                if (source == null || source instanceof NullSCMSource) {
+                    continue;
+                }
+                SCMHead oldHead = oldBranch.getHead();
+                SCMHead newHead = SCMHeadMigration.readResolveSCMHead(source, oldHead);
+                if (newHead != oldHead) {
+                    LOGGER.log(Level.INFO, "Job {0}: a plugin upgrade is requesting migration of branch from {1} to {2}",
+                            new Object[]{item.getFullName(), oldHead.getClass(), newHead.getClass()}
+                    );
+                    try {
+                        Branch newBranch = new Branch(oldBranch.getSourceId(), newHead, oldBranch.getScm(),
+                                oldBranch.getProperties());
+                        newBranch.setActions(oldBranch.getActions());
+                        factory.setBranch(item, newBranch);
+                        SCMRevision revision = factory.getRevision(item);
+                        factory.setRevisionHash(item, SCMHeadMigration.readResolveSCMRevision(source, revision));
```

Comment:
if you change the head you most likely need to change the revision too (as the revision includes the head)... thus if the head is changed we change the revision...

I guess we could modify the contract to resolve the revision too for future fix-up